src/nemos/solvers.py

```diff
@@ -0,0 +1,725 @@
+            For GLMs these are:
+                X : DESIGN_INPUT_TYPE
+                    Input data.
+                y : jnp.ndarray
+                    Output data.
+
+        Returns
+        -------
+        OptStep
+            xs :
+                Parameters after taking one step defined in the inner loop of Prox-SVRG.
+            state :
+                Updated state.
+        """
+        # NOTE this doesn't update state.xs, state.df_xs, that has to be done outside
+
+        next_params = self._xk_update_step(
+            current_params, state.xs, state.df_xs, state.stepsize, prox_lambda, *args
+        )
+
+        state = state._replace(
+            iter_num=state.iter_num + 1,
+        )
+
+        return OptStep(params=next_params, state=state)
+
+    @partial(jit, static_argnums=(0,))
+    def run(
+        self,
+        init_params: Pytree,
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> OptStep:
+        """
+        Run a whole optimization until convergence or until `maxiter` epochs are reached.
+        Called from `GLM.fit` and assumes that X and y are the full data set.
+
+        Parameters
+        ----------
+        init_params :
+            Initial parameters to start from.
+        prox_lambda :
+            Regularization strength.
+        args:
+            Positional arguments passed to loss function `fun` and its gradient (e.g. `fun(current_params, *args)`),
+            most likely input and output data.
+            They are expected to be Pytrees with arrays or FeaturePytree as their leaves, with all of their
+            leaves having the same sized first dimension (corresponding to the number of data points).
+            For GLMs these are:
+                X : DESIGN_INPUT_TYPE
+                    Input data.
+                y : jnp.ndarray
+                    Output data.
+
+        Returns
+        -------
+        OptStep
+            final_params :
+                Parameters at the end of the last innner loop.
+                (... or the average of the parameters over the last inner loop)
+            final_state :
+                Final optimizer state.
+        """
+        # initialize the state, including the full gradient at the initial parameters
+        init_state = self.init_state(
+            init_params,
+            prox_lambda,
+            *args,
+            init_full_gradient=True,
+        )
+
+        return self._run(init_params, init_state, prox_lambda, *args)
+
+    @partial(jit, static_argnums=(0,))
+    def _run(
+        self,
+        init_params: Pytree,
+        init_state: SVRGState,
+        prox_lambda: Union[float, None],
+        *args,
+    ) -> OptStep:
+        """
+        Run a whole optimization until convergence or until `maxiter` epochs are reached.
+        Called from `GLM.fit` and assumes that X and y are the full data set.
```

Comment:
same point about GLM being too specific

Reply:
Changed it. Let me know if you like it the new way.